File: src/composer.cpp
#include "hanfe/composer.hpp"

#include <unordered_map>
#include <unordered_set>

namespace hanfe {
namespace {

struct PairHash {
    size_t operator()(const std::pair<char32_t, char32_t>& pair) const noexcept {
        return static_cast<size_t>(pair.first) * 1315423911u + static_cast<size_t>(pair.second);
    }
};

const char32_t CHO_LIST[] = {U'ㄱ', U'ㄲ', U'ㄴ', U'ㄷ', U'ㄸ', U'ㄹ', U'ㅁ', U'ㅂ', U'ㅃ', U'ㅅ',
                             U'ㅆ', U'ㅇ', U'ㅈ', U'ㅉ', U'ㅊ', U'ㅋ', U'ㅌ', U'ㅍ', U'ㅎ'};

const char32_t JUNG_LIST[] = {U'ㅏ', U'ㅐ', U'ㅑ', U'ㅒ', U'ㅓ', U'ㅔ', U'ㅕ', U'ㅖ', U'ㅗ', U'ㅘ',
                              U'ㅙ', U'ㅚ', U'ㅛ', U'ㅜ', U'ㅝ', U'ㅞ', U'ㅟ', U'ㅠ', U'ㅡ', U'ㅢ',
                              U'ㅣ'};

const char32_t JONG_LIST[] = {U'\0', U'ㄱ', U'ㄲ', U'ㄳ', U'ㄴ', U'ㄵ', U'ㄶ', U'ㄷ', U'ㄹ', U'ㄺ',
                              U'ㄻ', U'ㄼ', U'ㄽ', U'ㄾ', U'ㄿ', U'ㅀ', U'ㅁ', U'ㅂ', U'ㅄ', U'ㅅ',
                              U'ㅆ', U'ㅇ', U'ㅈ', U'ㅊ', U'ㅋ', U'ㅌ', U'ㅍ', U'ㅎ'};

const std::unordered_map<std::pair<char32_t, char32_t>, char32_t, PairHash> DOUBLE_INITIAL = {
    {{U'ㄱ', U'ㄱ'}, U'ㄲ'}, {{U'ㄷ', U'ㄷ'}, U'ㄸ'}, {{U'ㅂ', U'ㅂ'}, U'ㅃ'},
    {{U'ㅈ', U'ㅈ'}, U'ㅉ'}, {{U'ㅅ', U'ㅅ'}, U'ㅆ'}};

const std::unordered_map<char32_t, std::pair<char32_t, char32_t>> INITIAL_DECOMPOSE = [] {
    std::unordered_map<char32_t, std::pair<char32_t, char32_t>> table;
    for (const auto& entry : DOUBLE_INITIAL) {
        table.emplace(entry.second, entry.first);
    }
    return table;
}();

const std::unordered_map<std::pair<char32_t, char32_t>, char32_t, PairHash> DOUBLE_MEDIAL = {
    {{U'ㅗ', U'ㅏ'}, U'ㅘ'}, {{U'ㅗ', U'ㅐ'}, U'ㅙ'}, {{U'ㅗ', U'ㅣ'}, U'ㅚ'},
    {{U'ㅜ', U'ㅓ'}, U'ㅝ'}, {{U'ㅜ', U'ㅔ'}, U'ㅞ'}, {{U'ㅜ', U'ㅣ'}, U'ㅟ'},
    {{U'ㅡ', U'ㅣ'}, U'ㅢ'}};

const std::unordered_map<char32_t, std::pair<char32_t, char32_t>> MEDIAL_DECOMPOSE = [] {
    std::unordered_map<char32_t, std::pair<char32_t, char32_t>> table;
    for (const auto& entry : DOUBLE_MEDIAL) {
        table.emplace(entry.second, entry.first);
    }
    return table;
}();

const std::unordered_map<std::pair<char32_t, char32_t>, char32_t, PairHash> DOUBLE_FINAL = {
    {{U'ㄱ', U'ㄱ'}, U'ㄲ'}, {{U'ㄱ', U'ㅅ'}, U'ㄳ'}, {{U'ㄴ', U'ㅈ'}, U'ㄵ'},
    {{U'ㄴ', U'ㅎ'}, U'ㄶ'}, {{U'ㄹ', U'ㄱ'}, U'ㄺ'}, {{U'ㄹ', U'ㅁ'}, U'ㄻ'},
    {{U'ㄹ', U'ㅂ'}, U'ㄼ'}, {{U'ㄹ', U'ㅅ'}, U'ㄽ'}, {{U'ㄹ', U'ㅌ'}, U'ㄾ'},
    {{U'ㄹ', U'ㅍ'}, U'ㄿ'}, {{U'ㄹ', U'ㅎ'}, U'ㅀ'}, {{U'ㅂ', U'ㅅ'}, U'ㅄ'},
    {{U'ㅅ', U'ㅅ'}, U'ㅆ'}};

const std::unordered_map<char32_t, std::pair<char32_t, char32_t>> FINAL_DECOMPOSE = [] {
    std::unordered_map<char32_t, std::pair<char32_t, char32_t>> table;
    for (const auto& entry : DOUBLE_FINAL) {
        table.emplace(entry.second, entry.first);
    }
    return table;
}();

const std::unordered_map<char32_t, int> CHOSEONG_INDEX = [] {
    std::unordered_map<char32_t, int> table;
    for (size_t idx = 0; idx < std::size(CHO_LIST); ++idx) {
        table.emplace(CHO_LIST[idx], static_cast<int>(idx));
    }
    return table;
}();

const std::unordered_map<char32_t, int> JUNGSEONG_INDEX = [] {
    std::unordered_map<char32_t, int> table;
    for (size_t idx = 0; idx < std::size(JUNG_LIST); ++idx) {
        table.emplace(JUNG_LIST[idx], static_cast<int>(idx));
    }
    return table;
}();

const std::unordered_map<char32_t, int> JONGSEONG_INDEX = [] {
    std::unordered_map<char32_t, int> table;
    for (size_t idx = 0; idx < std::size(JONG_LIST); ++idx) {
        table.emplace(JONG_LIST[idx], static_cast<int>(idx));
    }
    return table;
}();

const std::unordered_set<char32_t> CONSONANTS = [] {
    std::unordered_set<char32_t> set;
    for (char32_t ch : CHO_LIST) {
        set.insert(ch);
    }
    for (char32_t ch : JONG_LIST) {
        if (ch != U'\0') {
            set.insert(ch);
        }
    }
    return set;
}();

const std::unordered_set<char32_t> VOWELS = [] {
    std::unordered_set<char32_t> set;
    for (char32_t ch : JUNG_LIST) {
        set.insert(ch);
    }
    return set;
}();

bool is_consonant(char32_t ch) { return CONSONANTS.count(ch) > 0; }

bool is_vowel(char32_t ch) { return VOWELS.count(ch) > 0; }

}  // namespace

HangulComposer::HangulComposer() = default;

CompositionResult HangulComposer::feed(char32_t ch, JamoRole role) {
    std::u32string commit;
    if (is_vowel(ch)) {
        commit = handle_vowel(ch, role);
    } else {
        commit = handle_consonant(ch, role);
    }
    CompositionResult result;
    result.commit = utf8_from_u32string(commit);
    result.preedit = utf8_from_u32string(current_preedit());
    return result;
}

std::optional<std::string> HangulComposer::backspace() {
    if (trailing_) {
        auto it = FINAL_DECOMPOSE.find(*trailing_);
        if (it != FINAL_DECOMPOSE.end()) {
            trailing_ = it->second.first;
        } else {
            trailing_.reset();
        }
        return utf8_from_u32string(current_preedit());
    }
    if (vowel_) {
        auto it = MEDIAL_DECOMPOSE.find(*vowel_);
        if (it != MEDIAL_DECOMPOSE.end()) {
            vowel_ = it->second.first;
        } else {
            vowel_.reset();
            if (leading_ && *leading_ == U'ㅇ') {
                leading_.reset();
            }
        }
        return utf8_from_u32string(current_preedit());
    }
    if (leading_) {
        auto it = INITIAL_DECOMPOSE.find(*leading_);
        if (it != INITIAL_DECOMPOSE.end()) {
            leading_ = it->second.first;
        } else {
            leading_.reset();
        }
        return utf8_from_u32string(current_preedit());
    }
    return std::nullopt;
}

std::string HangulComposer::flush() {
    std::u32string commit = compose();
    leading_.reset();
    vowel_.reset();
    trailing_.reset();
    return utf8_from_u32string(commit);
}

std::u32string HangulComposer::handle_consonant(char32_t ch, JamoRole role) {
    std::u32string commit;
    bool force_trailing = role == JamoRole::Trailing;
    bool force_leading = role == JamoRole::Leading;

    if (!leading_) {
        leading_ = ch;
        trailing_.reset();
        return commit;
    }

    if (force_leading) {
        commit = compose();
        leading_ = ch;
        vowel_.reset();
        trailing_.reset();
        return commit;
    }

    if (!vowel_) {
        auto it = DOUBLE_INITIAL.find({*leading_, ch});
        if (it != DOUBLE_INITIAL.end()) {
            leading_ = it->second;
        } else {
            commit.push_back(*leading_);
            leading_ = ch;
        }
        return commit;
    }

    if (force_trailing) {
        return attach_trailing(ch);
    }

    if (!trailing_) {
        if (is_consonant(ch)) {
            trailing_ = ch;
            return commit;
        }
        commit = compose();
        leading_ = ch;
        vowel_.reset();
        trailing_.reset();
        return commit;
    }

    auto it = DOUBLE_FINAL.find({*trailing_, ch});
    if (it != DOUBLE_FINAL.end()) {
        trailing_ = it->second;
    } else {
        commit = compose();
        leading_ = ch;
        vowel_.reset();
        trailing_.reset();
    }
    return commit;
}

std::u32string HangulComposer::handle_vowel(char32_t ch, JamoRole /*role*/) {
    std::u32string commit;
    if (!leading_) {
        leading_ = U'ㅇ';
    }

    if (!vowel_) {
        vowel_ = ch;
        return commit;
    }

    auto it = DOUBLE_MEDIAL.find({*vowel_, ch});
    if (it != DOUBLE_MEDIAL.end()) {
        vowel_ = it->second;
        return commit;
    }

    if (trailing_) {
        auto split = FINAL_DECOMPOSE.find(*trailing_);
        if (split != FINAL_DECOMPOSE.end()) {
            auto [first, second] = split->second;
            trailing_ = first;
            commit = compose();
            leading_ = second;
            vowel_ = ch;
            trailing_.reset();
            return commit;
        }
        commit = compose();
        leading_ = U'ㅇ';
        vowel_ = ch;
        trailing_.reset();
        return commit;
    }

    commit = compose();
    leading_ = U'ㅇ';
    vowel_ = ch;
    trailing_.reset();
    return commit;
}

std::u32string HangulComposer::attach_trailing(char32_t ch) {
    std::u32string commit;
    if (!trailing_) {
        if (is_consonant(ch)) {
            trailing_ = ch;
            return commit;
        }
        commit = compose();
        leading_ = ch;
        vowel_.reset();
        trailing_.reset();
        return commit;
    }

    auto it = DOUBLE_FINAL.find({*trailing_, ch});
    if (it != DOUBLE_FINAL.end()) {
        trailing_ = it->second;
        return commit;
    }

    commit = compose();
    leading_ = ch;
    vowel_.reset();
    trailing_.reset();
    return commit;
}

std::u32string HangulComposer::compose() const {
    if (!leading_ && !vowel_) {
        return U"";
    }
    if (leading_ && vowel_) {
        int lead_index = CHOSEONG_INDEX.at(*leading_);
        int vowel_index = JUNGSEONG_INDEX.at(*vowel_);
        int tail_index = 0;
        if (trailing_) {
            tail_index = JONGSEONG_INDEX.at(*trailing_);
        }
        char32_t codepoint = static_cast<char32_t>(0xAC00 + ((lead_index * 21) + vowel_index) * 28 + tail_index);
        return std::u32string(1, codepoint);
    }
    if (leading_) {
        return std::u32string(1, *leading_);
    }
    if (vowel_) {
        return std::u32string(1, *vowel_);
    }
    return U"";
}

std::u32string HangulComposer::current_preedit() const { return compose(); }

}  // namespace hanfe

File: internal/hangul/composer.go
package hangul

import (
	"bytes"

	logotype "github.com/gg582/hangul-logotype/hangul"
)

type JamoRole int

const (
	RoleAuto JamoRole = iota
	RoleLeading
	RoleTrailing
)

type CompositionResult struct {
	Commit  string
	Preedit string
}

type HangulComposer struct {
	leading  []rune
	vowels   []rune
	trailing []rune
}

func NewHangulComposer() *HangulComposer {
	return &HangulComposer{
		leading:  make([]rune, 0, 2),
		vowels:   make([]rune, 0, 3),
		trailing: make([]rune, 0, 2),
	}
}

var (
	choList  = []rune{'ㄱ', 'ㄲ', 'ㄴ', 'ㄷ', 'ㄸ', 'ㄹ', 'ㅁ', 'ㅂ', 'ㅃ', 'ㅅ', 'ㅆ', 'ㅇ', 'ㅈ', 'ㅉ', 'ㅊ', 'ㅋ', 'ㅌ', 'ㅍ', 'ㅎ'}
	jungList = []rune{'ㅏ', 'ㅐ', 'ㅑ', 'ㅒ', 'ㅓ', 'ㅔ', 'ㅕ', 'ㅖ', 'ㅗ', 'ㅘ', 'ㅙ', 'ㅚ', 'ㅛ', 'ㅜ', 'ㅝ', 'ㅞ', 'ㅟ', 'ㅠ', 'ㅡ', 'ㅢ', 'ㅣ'}
	jongList = []rune{0, 'ㄱ', 'ㄲ', 'ㄳ', 'ㄴ', 'ㄵ', 'ㄶ', 'ㄷ', 'ㄹ', 'ㄺ', 'ㄻ', 'ㄼ', 'ㄽ', 'ㄾ', 'ㄿ', 'ㅀ', 'ㅁ', 'ㅂ', 'ㅄ', 'ㅅ', 'ㅆ', 'ㅇ', 'ㅈ', 'ㅊ', 'ㅋ', 'ㅌ', 'ㅍ', 'ㅎ'}
)

var doubleInitial = map[[2]rune]rune{
	{'ㄱ', 'ㄱ'}: 'ㄲ',
	{'ㄷ', 'ㄷ'}: 'ㄸ',
	{'ㅂ', 'ㅂ'}: 'ㅃ',
	{'ㅈ', 'ㅈ'}: 'ㅉ',
	{'ㅅ', 'ㅅ'}: 'ㅆ',
}

var (
	consonantSet = buildSet(append(append([]rune{}, choList...), filterZero(jongList)...))
	vowelSet     = buildSet(jungList)
	finalSet     = buildSet(filterZero(jongList))
)

func buildSet(list []rune) map[rune]struct{} {
	out := make(map[rune]struct{}, len(list))
	for _, r := range list {
		out[r] = struct{}{}
	}
	return out
}

func filterZero(list []rune) []rune {
	var out []rune
	for _, r := range list {
		if r != 0 {
			out = append(out, r)
		}
	}
	return out
}

func isConsonant(r rune) bool {
	_, ok := consonantSet[r]
	return ok
}

func isVowel(r rune) bool {
	_, ok := vowelSet[r]
	return ok
}

func (c *HangulComposer) Feed(ch rune, role JamoRole) CompositionResult {
	var commit []rune
	if isVowel(ch) {
		commit = c.handleVowel(ch)
	} else {
		commit = c.handleConsonant(ch, role)
	}
	return CompositionResult{
		Commit:  string(commit),
		Preedit: string(c.currentPreedit()),
	}
}

func (c *HangulComposer) Backspace() (string, bool) {
	switch {
	case len(c.trailing) > 0:
		c.trailing = c.trailing[:len(c.trailing)-1]
	case len(c.vowels) > 0:
		c.vowels = c.vowels[:len(c.vowels)-1]
	case len(c.leading) > 0:
		c.leading = c.leading[:len(c.leading)-1]
	default:
		return "", false
	}
	return string(c.currentPreedit()), true
}

func (c *HangulComposer) Flush() string {
	commit := string(c.composeCurrent())
	c.reset()
	return commit
}

func (c *HangulComposer) handleConsonant(ch rune, role JamoRole) []rune {
	var commit []rune

	if !isConsonant(ch) {
		commit = append(commit, c.composeCurrent()...)
		c.reset()
		return append(commit, ch)
	}

	if len(c.leading) == 0 {
		if len(c.vowels) > 0 {
			commit = append(commit, c.composeCurrent()...)
			c.reset()
		}
		c.leading = append(c.leading[:0], ch)
		c.trailing = c.trailing[:0]
		return commit
	}

	if role == RoleLeading {
		commit = append(commit, c.composeCurrent()...)
		c.reset()
		c.leading = append(c.leading[:0], ch)
		return commit
	}

	if len(c.vowels) == 0 {
		if len(c.leading) == 1 {
			pair := [2]rune{c.leading[0], ch}
			if _, ok := doubleInitial[pair]; ok {
				c.leading = append(c.leading, ch)
				return commit
			}
		}
		commit = append(commit, composeLeading(c.leading)...)
		c.leading = append(c.leading[:0], ch)
		c.trailing = c.trailing[:0]
		return commit
	}

	if role == RoleTrailing {
		return c.attachTrailing(ch)
	}

	if len(c.trailing) == 0 {
		c.trailing = append(c.trailing[:0], ch)
		return commit
	}

	candidate := append(append([]rune{}, c.trailing...), ch)
	if isValidFinal(candidate) {
		c.trailing = candidate
		return commit
	}

	commit = append(commit, c.composeCurrent()...)
	c.reset()
	c.leading = append(c.leading[:0], ch)
	return commit
}

func (c *HangulComposer) handleVowel(ch rune) []rune {
	var commit []rune

	if len(c.trailing) > 0 {
		commit = append(commit, c.detachTrailingForLeading()...)
	}

	if len(c.vowels) == 0 {
		c.vowels = append(c.vowels[:0], ch)
		return commit
	}

	candidate := append(append([]rune{}, c.vowels...), ch)
	if isCombinedVowel(candidate) {
		c.vowels = append(c.vowels, ch)
		return commit
	}

	if len(c.leading) == 0 {
		commit = append(commit, composeRunes(c.vowels)...)
	} else {
		commit = append(commit, c.composeCurrent()...)
	}

	c.leading = c.leading[:0]
	c.vowels = append(c.vowels[:0], ch)
	c.trailing = c.trailing[:0]
	return commit
}

func (c *HangulComposer) attachTrailing(ch rune) []rune {
	var commit []rune

	if !isConsonant(ch) {
		commit = append(commit, c.composeCurrent()...)
		c.reset()
		c.leading = append(c.leading[:0], ch)
		return commit
	}

	if len(c.trailing) == 0 {
		c.trailing = append(c.trailing[:0], ch)
		return commit
	}

	candidate := append(append([]rune{}, c.trailing...), ch)
	if isValidFinal(candidate) {
		c.trailing = candidate
		return commit
	}

	commit = append(commit, c.composeCurrent()...)
	c.reset()
	c.leading = append(c.leading[:0], ch)
	return commit
}

func (c *HangulComposer) detachTrailingForLeading() []rune {
	carry := c.trailing
	nextLeading := carry[len(carry)-1]
	if len(carry) > 1 {
		c.trailing = carry[:len(carry)-1]
	} else {
		c.trailing = c.trailing[:0]
	}
	committed := c.composeCurrent()
	c.leading = append(c.leading[:0], nextLeading)
	c.vowels = c.vowels[:0]
	c.trailing = c.trailing[:0]
	return committed
}

func (c *HangulComposer) composeCurrent() []rune {
	raw := make([]rune, 0, 1+len(c.vowels)+len(c.trailing))
	if lead := composeLeading(c.leading); len(lead) > 0 {
		raw = append(raw, lead...)
	}
	raw = append(raw, c.vowels...)
	raw = append(raw, c.trailing...)
	return composeRunes(raw)
}

func (c *HangulComposer) currentPreedit() []rune {
	return c.composeCurrent()
}

func (c *HangulComposer) reset() {
	c.leading = c.leading[:0]
	c.vowels = c.vowels[:0]
	c.trailing = c.trailing[:0]
}

func composeLeading(parts []rune) []rune {
	switch len(parts) {
	case 0:
		return nil
	case 1:
		return []rune{parts[0]}
	case 2:
		pair := [2]rune{parts[0], parts[1]}
		if combined, ok := doubleInitial[pair]; ok {
			return []rune{combined}
		}
		return []rune{parts[0], parts[1]}
	default:
		return append([]rune{}, parts...)
	}
}

func isCombinedVowel(candidate []rune) bool {
	composed := composeRunes(candidate)
	return len(composed) == 1 && isVowel(composed[0])
}

func isValidFinal(candidate []rune) bool {
	composed := composeRunes(candidate)
	if len(composed) != 1 {
		return false
	}
	_, ok := finalSet[composed[0]]
	return ok
}

func composeRunes(raw []rune) []rune {
	if len(raw) == 0 {
		return nil
	}
	var buf bytes.Buffer
	logotype.LogoType(&buf, raw)
	return []rune(buf.String())
}
